EasyPost/Utilities/Cryptography.cs

```diff
@@ -0,0 +1,152 @@
+
+        /// <summary>
+        ///     Convert a byte array to a hex string.
+        /// </summary>
+        /// <param name="bytes">Byte array to convert to hex string.</param>
+        /// <returns>Hex string</returns>
+        private static string AsHexString(this byte[] bytes)
+        {
+            // Fastest safe way to convert a byte array to hex string,
+            // per https://stackoverflow.com/a/624379/13343799
+
+            uint[] lookup32 = Lookup32;
+            char[] result = new char[bytes.Length * 2];
+            for (int i = 0; i < bytes.Length; i++)
+            {
+                uint val = lookup32[bytes[i]];
+                result[2 * i] = (char)val;
+                result[2 * i + 1] = (char)(val >> 16);
+            }
+
+            return new string(result).ToLower();
+        }
+
+        /// <summary>
+        ///     Convert a string to a hex string using a specific encoding (defaults to UTF-8)
+        /// </summary>
+        /// <param name="str">String to convert to hex string.</param>
+        /// <param name="encoding">Encoding to use. Default: UTF-8</param>
+        /// <returns>Hex string</returns>
+        public static string AsHexString(this string str, Encoding? encoding = null)
+        {
+            byte[] bytes = str.AsByteArray(encoding);
+
+            return bytes.AsHexString();
+        }
+
+        /// <summary>
+        ///     Convert a byte array to a string using a specific encoding (defaults to UTF-8)
+        /// </summary>
+        /// <param name="bytes">Byte array to convert to string.</param>
+        /// <param name="encoding">Encoding to use. Default: UTF-8</param>
+        /// <returns>String</returns>
+        public static string AsString(this byte[] bytes, Encoding? encoding = null)
+        {
+            encoding ??= Encoding.UTF8;
+
+            return encoding.GetString(bytes);
+        }
+
+        /// <summary>
+        ///     Calculate the HMAC-SHA256 hex digest of a byte array.
+        /// </summary>
+        /// <param name="data">Data to calculate hex digest for.</param>
+        /// <param name="secret">Key used to calculate data hex digest.</param>
+        /// <param name="normalizationForm">Normalization type to use when normalizing key. Default: No normalization.</param>
+        /// <returns>Hex digest of data.</returns>
+        public static string CalculateHMACSHA256HexDigest(this byte[] data, string secret, NormalizationForm? normalizationForm = null)
+        {
+            if (normalizationForm != null)
+            {
+                secret = secret.Normalize(normalizationForm.Value);
+            }
+
+            byte[] keyBytes = Encoding.UTF8.GetBytes(secret);
+
+            using HMACSHA256 hmac = new HMACSHA256(keyBytes);
+            byte[] hash = hmac.ComputeHash(data);
+            return hash.AsHexString();
+        }
+
+
+        /// <summary>
+        ///     Check whether two signatures match. This is safe against timing attacks.
```

Comment:
Question: How are these safe against timing attacks? You are still just comparing strings which isn't typically safe.

We should probably implement something like [SlowEquals](https://gist.github.com/adrianstevens/b053ec17cfc5ca868e71) to assist with this.

Reply:
It's safe against timing attacks because it does not return early when a mismatch is detected. It waits to go through the entire array before returning a boolean, making it impossible to tell where in the array the mismatch occurred.

Reply:
Can you provide some documentation on the matter? How do we know this?

Reply:
https://codahale.com/a-lesson-in-timing-attacks/

Reply:
I was looking at some weird cached results for this page and the initial commit. It appears my concerns have since been addressed.